File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/InitialPasswordControllerTest.java
package uk.gov.justice.digital.hmpps.oauth2server.resource;

import com.microsoft.applicationinsights.TelemetryClient;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserEmail;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken;
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken.TokenType;
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.AccountDetail;
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.StaffUserAccount;
import uk.gov.justice.digital.hmpps.oauth2server.security.UserService;
import uk.gov.justice.digital.hmpps.oauth2server.verify.ResetPasswordService;
import uk.gov.justice.digital.hmpps.oauth2server.verify.TokenService;

import javax.servlet.http.HttpServletRequest;
import java.util.Optional;
import java.util.Set;

import static java.util.Map.entry;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class InitialPasswordControllerTest {
    @Mock
    private ResetPasswordService resetPasswordService;
    @Mock
    private TokenService tokenService;
    @Mock
    private UserService userService;
    @Mock
    private TelemetryClient telemetryClient;
    @Mock
    private HttpServletRequest request;

    private InitialPasswordController controller;

    @Before
    public void setUp() {
        controller = new InitialPasswordController(resetPasswordService, tokenService, userService, telemetryClient, Set.of("password1"), "http://hdcurl");
    }

    @Test
    public void initialPasswordSuccess_checkView() {
        final var modelAndView = controller.initialPasswordSuccess(request);
        assertThat(modelAndView.getViewName()).isEqualTo("initialPasswordSuccess");
    }

    @Test
    public void initialPasswordSuccess_checkModel() {
        final var modelAndView = controller.initialPasswordSuccess(request);
        assertThat(modelAndView.getModel()).containsOnly(entry("hdcUrl", "http://hdcurl"));
    }

    @Test
    public void initialPasswordSuccess_hdcNotSet() {
        when(request.getContextPath()).thenReturn("/context");
        final var modelAndView = new InitialPasswordController(resetPasswordService, tokenService, userService, telemetryClient, Set.of("password1"), null).initialPasswordSuccess(request);
        assertThat(modelAndView.getModel()).containsOnly(entry("hdcUrl", "/context/login"));
    }

    @Test
    public void initialPassword_checkView() {
        setupCheckTokenValid();
        final var modelAndView = controller.initialPassword("token");
        assertThat(modelAndView.getViewName()).isEqualTo("setPassword");
    }

    @Test
    public void initialPassword_checkModel() {
        setupCheckTokenValid();
        final var modelAndView = controller.initialPassword("sometoken");
        assertThat(modelAndView.getModel()).containsOnly(entry("token", "sometoken"), entry("isAdmin", Boolean.FALSE), entry("initial", Boolean.TRUE));
    }

    @Test
    public void initialPassword_checkModelAdminUser() {
        final var user = setupGetUserCallForProfile();
        user.getAccountDetail().setProfile("TAG_ADMIN");
        setupCheckAndGetTokenValid();
        final var modelAndView = controller.initialPassword("sometoken");
        assertThat(modelAndView.getModel()).containsOnly(entry("token", "sometoken"), entry("isAdmin", Boolean.TRUE), entry("initial", Boolean.TRUE));
    }

    @Test
    public void initialPassword_FailureCheckView() {
        when(tokenService.checkToken(any(), anyString())).thenReturn(Optional.of("invalid"));
        final var modelAndView = controller.initialPassword("token");
        assertThat(modelAndView.getViewName()).isEqualTo("resetPassword");
    }

    @Test
    public void initialPassword_FailureCheckModel() {
        when(tokenService.checkToken(any(), anyString())).thenReturn(Optional.of("expired"));
        final var modelAndView = controller.initialPassword("sometoken");
        assertThat(modelAndView.getModel()).containsOnly(entry("error", "expired"));
    }

    private void setupCheckTokenValid() {
        when(tokenService.checkToken(any(), anyString())).thenReturn(Optional.empty());
    }

    private void setupCheckAndGetTokenValid() {
        when(tokenService.checkToken(any(), anyString())).thenReturn(Optional.empty());
        when(tokenService.getToken(any(), anyString())).thenReturn(Optional.of(new UserToken(TokenType.RESET, UserEmail.of("user"))));
    }

    private StaffUserAccount setupGetUserCallForProfile() {
        final var user = new StaffUserAccount();
        user.setAccountDetail(new AccountDetail());
        when(userService.findUser(anyString())).thenReturn(Optional.of(user));
        return user;
    }
}

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/resource/InitialPasswordControllerTest.kt
package uk.gov.justice.digital.hmpps.oauth2server.resource

import com.microsoft.applicationinsights.TelemetryClient
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.entry
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.User
import uk.gov.justice.digital.hmpps.oauth2server.auth.model.UserToken
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.AccountDetail
import uk.gov.justice.digital.hmpps.oauth2server.nomis.model.NomisUserPersonDetails
import uk.gov.justice.digital.hmpps.oauth2server.security.UserService
import uk.gov.justice.digital.hmpps.oauth2server.verify.InitialPasswordService
import uk.gov.justice.digital.hmpps.oauth2server.verify.ResetPasswordService
import uk.gov.justice.digital.hmpps.oauth2server.verify.TokenService
import java.util.Optional
import javax.servlet.http.HttpServletRequest

class InitialPasswordControllerTest {
  private val resetPasswordService: ResetPasswordService = mock()
  private val initialPasswordService: InitialPasswordService = mock()
  private val tokenService: TokenService = mock()
  private val userService: UserService = mock()
  private val telemetryClient: TelemetryClient = mock()
  private val request: HttpServletRequest = mock()
  private val controller = InitialPasswordController(
    resetPasswordService,
    initialPasswordService,
    tokenService,
    userService,
    telemetryClient,
    setOf("password1"),
    true
  )

  @Nested
  inner class InitialPasswordSuccess {
    @Test
    fun initialPasswordSuccess_checkView() {
      val modelAndView = controller.initialPasswordSuccess()
      assertThat(modelAndView).isEqualTo("initialPasswordSuccess")
    }
  }

  @Nested
  inner class InitialPassword {
    @Test
    fun `initialPassword no token`() {
      val modelAndView = controller.initialPassword("  ", request)
      assertThat(modelAndView.viewName).isEqualTo("redirect:/reset-password")
      verifyZeroInteractions(tokenService)
    }

    @Test
    fun initialPassword_checkView() {
      setupGetUserCallForProfile()
      setupCheckAndGetTokenValid()
      val modelAndView = controller.initialPassword("token", request)
      assertThat(modelAndView.viewName).isEqualTo("setPassword")
    }

    @Test
    fun initialPassword_checkModel() {
      setupCheckAndGetTokenValid()
      setupGetUserCallForProfile()
      val modelAndView = controller.initialPassword("sometoken", request)
      assertThat(modelAndView.model).containsExactlyInAnyOrderEntriesOf(
        mapOf(
          "token" to "sometoken",
          "isAdmin" to false,
          "initial" to true,
          "username" to "someuser"
        )
      )
    }

    @Test
    fun initialPassword_checkModelAdminUser() {
      val user = setupGetUserCallForProfile()
      user.accountDetail.profile = "TAG_ADMIN"
      setupCheckAndGetTokenValid()
      val modelAndView = controller.initialPassword("sometoken", request)
      assertThat(modelAndView.model).containsExactlyInAnyOrderEntriesOf(
        mapOf(
          "token" to "sometoken",
          "isAdmin" to true,
          "initial" to true,
          "username" to "someuser"
        )
      )
    }

    @Test
    fun initialPassword_FailureExpiredCheckView() {
      whenever(request.requestURL).thenReturn(StringBuffer("someurl"))
      whenever(tokenService.checkToken(any(), anyString())).thenReturn(Optional.of("expired"))
      val modelAndView = controller.initialPassword("token", request)
      assertThat(modelAndView.viewName).isEqualTo("redirect:/initial-password-expired")
    }

    @Test
    fun initialPassword_FailureInvalidCheckView() {
      whenever(request.requestURL).thenReturn(StringBuffer("someurl"))
      whenever(tokenService.checkToken(any(), anyString())).thenReturn(Optional.of("invalid"))
      val modelAndView = controller.initialPassword("token", request)
      assertThat(modelAndView.viewName).isEqualTo("redirect:/reset-password")
    }

    @Test
    fun initialPassword_FailureCheckModel() {
      whenever(request.requestURL).thenReturn(StringBuffer("someurl"))
      whenever(tokenService.checkToken(any(), anyString())).thenReturn(Optional.of("expired"))
      val modelAndView = controller.initialPassword("sometoken", request)
      assertThat(modelAndView.model).containsOnly(entry("token", "sometoken"))
    }
  }

  @Nested
  inner class InitialPasswordLinkExpired {
    @Test
    fun initialPasswordLinkExpired_checkView() {
      whenever(request.requestURL).thenReturn(StringBuffer("someurl"))
      whenever(tokenService.checkToken(any(), anyString())).thenReturn(Optional.of("expired"))
      whenever(tokenService.getUserFromToken(any(), anyString())).thenReturn(User.builder().username("bob").build())
      whenever(initialPasswordService.resendInitialPasswordLink(anyString(), anyString())).thenReturn("newToken")
      val modelAndView = controller.initialPasswordLinkExpired("sometoken", request)
      assertThat(modelAndView.viewName).isEqualTo("createPasswordExpired")
    }

    @Test
    fun initialPasswordLinkExpired_checkModel() {
      whenever(request.requestURL).thenReturn(StringBuffer("someurl"))
      whenever(tokenService.checkToken(any(), anyString())).thenReturn(Optional.of("expired"))
      whenever(tokenService.getUserFromToken(any(), anyString())).thenReturn(User.builder().username("bob").build())
      whenever(initialPasswordService.resendInitialPasswordLink(anyString(), anyString())).thenReturn("newToken")
      val modelAndView = controller.initialPasswordLinkExpired("sometoken", request)
      assertThat(modelAndView.model).containsOnly(entry("link", "newToken"))
    }
  }

  private fun setupCheckAndGetTokenValid() {
    whenever(tokenService.checkToken(any(), anyString())).thenReturn(Optional.empty())
    whenever(tokenService.getToken(any(), anyString())).thenReturn(
      Optional.of(
        User.of("someuser").createToken(UserToken.TokenType.RESET)
      )
    )
  }

  private fun setupGetUserCallForProfile(): NomisUserPersonDetails {
    val user = NomisUserPersonDetails()
    user.accountDetail = AccountDetail()
    whenever(userService.findMasterUserPersonDetails(anyString())).thenReturn(Optional.of(user))
    return user
  }
}
